### PR TITLE
Helper to render within another layout

### PR DESCRIPTION
**Pull request by lowjoel**
This allows us to nest a layout within another, useful for all the course components where we have a sidebar within the global application template.

I think I need to find a way to test this.

For discussion: maybe `render within_layout: 'layout'` is better?

### CODE REVIEW

File: lib/extensions/action_controller/base.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  # Gets the layout hierarchy, from the outermost to the innermost layout.
+  #
+  # @return [Array] The layout hierarchy for this controller.
+  def layout_hierarchy
+    @layout_hierarchy ||=
+      Extensions::ActionController::Base.class_hierarchy(self.class).
+        select { |klass| klass < ActionController::Base }.
+        map { |klass| Extensions::ActionController::Base.class_layout(klass, self) }.
+        select { |layout| !layout.nil? }.
+        reduce([]) do |memo, layout|
```
Comment (lowjoel) [inline lib/extensions/action_controller/base.rb]: @jsyeo ...not again.

I personally prefer fold_left/fold_right, but meh.

**Comment by lowjoel** [comment]
Why did coverage go down >.<

**Comment by lowjoel** [comment]
Oh one test failed.

File: lib/extensions/action_controller/base.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    _layout
+  end
+
+  # Gets the parent layout of the given layout, as specified in the layout hierarchy.
+  #
+  # @param layout [Symbol|nil] The layout to obtain the parent of. If this is nil, obtains the
+  #                            current controller's parent layout.
+  # @return [Symbol] The parent layout of the given layout.
+  def parent_layout(layout = nil)
+    layout ||= current_layout
```
Comment (lowjoel) [inline lib/extensions/action_controller/base.rb]: maybe I should go:
```ruby
controller.parent_layout
```
to give the current controller's parent layout

and

```ruby
controller.parent_layout(of_layout: 'layout')
```
what do you think?

Comment (wangqiang1208) [inline lib/extensions/action_controller/base.rb]: Yeah, you new solution looks less confusion.

**Pull request merged**